The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5243343441/jobs/9467911685) [(plain)](https://github.com/rust-lang-ci/rust/commit/8a6082a639bc4f38df2479d3e9489402bd1e3d35/checks/14186143362/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  'TOOLSTATE_PUBLISH': '1',
  'TOOLSTATE_REPO': 'https://github.com/rust-lang-nursery/rust-toolstate',
  'TOOLSTATE_REPO_ACCESS_TOKEN': '***',
  '_': '/usr/bin/python3'}
stage-build INFO: Skipping building of unimportant components for a try build
stage-build DEBUG: Changing working dir from `/checkout/obj` to `/tmp/tmp-multistage/opt-artifacts/rustc-perf`
stage-build INFO: Executing `RUSTC=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc RUSTC_BOOTSTRAP=1 /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build -p collector`
    Updating crates.io index
    Updating git repository `https://github.com/rust-lang/measureme`
---
   --> library/core/src/slice/iter/macros.rs:195:27
    |
47  | /  macro_rules! iterator {
48  | |      (
49  | |          struct $name:ident -> $ptr:ty,
50  | |          $elem:ty,
...   |
195 | |              fn fold<B, F>(mut self, init: B, mut f: F) -> B
    | |                            |
    | |                            help: remove this `mut`
...   |
428 | |      }
428 | |      }
429 | |  }
    | |__- in this expansion of `iterator!`
    |
   ::: library/core/src/slice/iter.rs:132:1
    |
132 |  / iterator! {struct Iter -> *const T, &'a T, const, {/* no mut */}, {
133 |  |     fn is_sorted_by<F>(self, mut compare: F) -> bool
134 |  |     where
135 |  |         Self: Sized,
139 |  |     }
140 |  | }}
    |  |__- in this macro invocation
    |
    |
    = note: `-D unused-mut` implied by `-D warnings`
error: variable does not need to be mutable
   --> library/core/src/slice/iter/macros.rs:195:27
    |
47  | / macro_rules! iterator {
47  | / macro_rules! iterator {
48  | |     (
49  | |         struct $name:ident -> $ptr:ty,
50  | |         $elem:ty,
...   |
195 | |             fn fold<B, F>(mut self, init: B, mut f: F) -> B
    | |                           |
    | |                           help: remove this `mut`
...   |
428 | |     }
428 | |     }
429 | | }
    | |_- in this expansion of `iterator!`
    |
   ::: library/core/src/slice/iter.rs:364:1
    |
364 |   iterator! {struct IterMut -> *mut T, &'a mut T, mut, {mut}, {}}

[RUSTC-TIMING] core test:false 6.605
error: could not compile `core` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
---
Total duration:                                               10s
-----------------------------------------------------------------
root INFO: Free disk space: 512.16 GiB out of total 581.32 GiB (11.89% used)
Traceback (most recent call last):
  File "../src/ci/stage-build.py", line 1030, in <module>
    run(runner)
  File "../src/ci/stage-build.py", line 1020, in run
    raise e
  File "../src/ci/stage-build.py", line 1017, in run
    execute_build_pipeline(timer, pipeline, runner, build_args)
  File "../src/ci/stage-build.py", line 897, in execute_build_pipeline
    pipeline.rustc_profile_dir_root()
  File "../src/ci/stage-build.py", line 642, in bootstrap_build
    cmd(arguments, env=env)
  File "../src/ci/stage-build.py", line 476, in cmd
    return subprocess.run(args, env=environment, check=True)
  File "/usr/lib64/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/usr/bin/python3', '/checkout/x.py', 'build', '--target', 'x86_64-unknown-linux-gnu', '--host', 'x86_64-unknown-linux-gnu', '--stage', '2', 'library/std', '--rust-profile-generate', '/tmp/tmp-multistage/opt-artifacts/rustc-pgo']' returned non-zero exit status 1.
```

</details>

--- a/library/core/src/slice/iter/macros.rs
+++ b/library/core/src/slice/iter/macros.rs
@@ -192,23 +192,33 @@ macro_rules! iterator {
             }
 
             #[inline]
-            fn fold<B, F>(mut self, init: B, mut f: F) -> B
+            fn fold<B, F>(self, init: B, mut f: F) -> B
                 where
                     F: FnMut(B, Self::Item) -> B,
             {
-                // Handling the 0-len case explicitly and then using a do-while style loop
-                // helps the optimizer. See issue #106288
+                // this implementation consists of the following optimizations compared to the
+                // default implementation:
+                // - do-while loop, as is llvm's preferred loop shape,
+                //   see https://releases.llvm.org/16.0.0/docs/LoopTerminology.html#more-canonical-loops
+                // - bumps an index instead of a pointer since the latter case inhibits
+                //   some optimizations, see #111603
+                // - avoids Option wrapping/matching
                 if is_empty!(self) {
                     return init;
                 }
                 let mut acc = init;
-                // SAFETY: The 0-len case was handled above so one loop iteration is guaranteed.
-                unsafe {
-                    loop {
-                        acc = f(acc, next_unchecked!(self));
-                        if is_empty!(self) {
-                            break;
-                        }
+                let mut i = 0;
+                let len = len!(self);
+                loop {
+                    // SAFETY: the loop iterates `i in 0..len`, which always is in bounds of
+                    // the slice allocation
+                    acc = f(acc, unsafe { & $( $mut_ )? *self.ptr.add(i).as_ptr() });
+                    // SAFETY: `i` can't overflow since it'll only reach usize::MAX if the
+                    // slice had that length, in which case we'll break out of the loop
+                    // after the increment
+                    i = unsafe { i.unchecked_add(1) };
+                    if i == len {
+                        break;
                     }
                 }
                 acc

--- a/tests/codegen/slice-iter-fold.rs
+++ b/tests/codegen/slice-iter-fold.rs
@@ -1,0 +1,14 @@
+// ignore-debug: the debug assertions get in the way
+// compile-flags: -O
+// min-llvm-version: 16
+#![crate_type = "lib"]
+
+// CHECK-LABEL: @slice_fold_to_last
+#[no_mangle]
+pub fn slice_fold_to_last(slice: &[i32]) -> Option<&i32> {
+    // CHECK-NOT: loop
+    // CHECK-NOT: br
+    // CHECK-NOT: call
+    // CHECK: ret
+    slice.iter().fold(None, |_, i| Some(i))
+}

--- a/tests/codegen/vec-shrink-panik.rs
+++ b/tests/codegen/vec-shrink-panik.rs
@@ -38,14 +38,6 @@ pub fn issue71861(vec: Vec<u32>) -> Box<[u32]> {
 #[no_mangle]
 pub fn issue75636<'a>(iter: &[&'a str]) -> Box<[&'a str]> {
     // CHECK-NOT: panic
-
-    // Call to panic_cannot_unwind in case of double-panic is expected,
-    // on LLVM 16 and older, but other panics are not.
-    // old: filter
-    // old-NEXT: ; call core::panicking::panic_cannot_unwind
-    // old-NEXT: panic_cannot_unwind
-
-    // CHECK-NOT: panic
     iter.iter().copied().collect()
 }
 

